Support for C wrapper to enable other languages to call your library.

edancain opened the pull request:
Hi Luc,
can't begin to thank you enough for this work, its impressive. 

I am not a RUST developer, and have to learn quickly over the past couple of weeks. I have made editions to your code and changes where necessary to support implementation of the wrapper. 

If you read the new readme.md ( /dji-log-parser/readme.md) there is an explanation as to what I have added to fork I made of your work. I have had to change a few use statements in files to support use serde::{Serialize, Deserialize};

Anyway, this now works for my purposes and I wanted to share back. 

Thanks again,
Edan

edancain committed:
testing running this unix exe from GO

edancain committed:
Go code to call the dji-log exe

edancain committed:
comments

edancain committed:
Initial Commit

edancain committed:
Attempts to make the static Lib compile correctly and parse the data.

edancain committed:
Implementation of a C wrapper around the Rust project to enable Go code integration, calls to the static Rust parsing library. Woot woot

edancain committed:
comments for clarity on how the C wrapper works. Name change to the Go file implementation

edancain committed:
Extensions to the wrapper to return a geojson string rather than writing it to disk, and io.Read read object in Go support.

edancain committed:
Comment change

edancain committed:
Learning as I go, dji-log-parser.h wasn't being push to the repo.

edancain committed:
Removing wrapper C functions that are unneeded. The functionality required is simply to pass back the string pointer of the geojson created from the file parsing.
Renamed free_string for clarity and to avoid conflicts. Adjusted the get_last_error to work with the new structure. Ensured all public C API functions marked with #[no_mangle]

Moved teh C API specific functions into the c_api.rs file to keep it all clean.

edancain committed:
I think I still have an issue but this code is working. The issue is when building the lib, I have set it to build the .h file automatically and this breaks implementation in GO. Need to work this out. But for now, this works

edancain committed:
Readme explanation of code implementation for C bindings.

edancain committed:
git ignoring local dev testing

edancain committed:
updates and changes to gitignore

edancain committed:
removing tests for go code

edancain committed:
updates to gitignore

edancain committed:
Delete .vscode directory

Tests for go code

edancain committed:
Delete test-data directory

test data for go

edancain committed:
Delete go-code directory

testing with go the C wrapper

edancain committed:
new code to update the automation of the C header file, ensuring that the file is auto generated and exposes the correct functions from both c_api.rs and lib.rs

edancain committed:
removing all go test files, adding them to gitignore

edancain committed:
changes to ignore go files when testing

edancain committed:
Update Cargo.toml

updates from main

edancain committed:
Update Cargo.lock

merging from main

edancain committed:
Update Cargo.toml

merging from main into master

edancain committed:
Add files via upload

merging from main into master

edancain committed:
Update Cargo.toml

merging main into master

edancain committed:
Update main.rs

merging main into master

edancain committed:
Update frame.rs

merging main into master

edancain committed:
Update request.rs

merging main into master

edancain committed:
Update auxiliary.rs

merging main into master

edancain committed:
Update details.rs

merging main into master

edancain committed:
Update lib.rs

merging main into master

edancain committed:
Update camera.rs

merging main into master

edancain committed:
Update gimbal.rs

merging main into master

edancain committed:
Update home.rs

merging main into master

edancain committed:
Update osd.rs

merging main into master

edancain committed:
Add files via upload

merging main with master

edancain committed:
update to gitignore

edancain committed:
merge from main

edancain committed:
merge from main

edancain committed:
merging from main

edancain committed:
C wrapper header file

edancain committed:
go mod, tried to ignore it but here it is

edancain committed:
merge from main

edancain committed:
Merge branch 'main'

edancain committed:
Delete go.mod

unneeded

edancain committed:
Quick change

lvauvillier commented:
Thank you very much for this PR and your kind words. I’m also learning Rust, and this is my first project :)

I quickly reviewed your changes and have a few initial comments / questions :

- I think the right way is to create a new crate at the root of the project (e.g., `dji-log-parser-c`). You can check, I have added JS bindings on a wasm target in a separate crate a few days ago.
- I see you added an option to the CLI to export keychains; this is definitely something I had in mind and planned to expose. I think it's better to place this code as a new Exporter. But this doesn't seem related to the C bindings, right?
- Why did you need to add `Deserialize`?
- Why did you need to change the `AuxiliaryInfo` struct?

edancain commented:
Hi Luc,
you are learning Rust? I would say you have mastered it. Me, this project of yours is the very first time I have ever used Rust, so my assumptions and code changes with regards to serde::{Serialize, Deserialize} and the AuxiliaryInfo struct were to fix build warnings really. 

With Deserialize, I wanted to add symmetry to data handling and I wasn't sure at the start if I would need it to convert data back to into a Rust struct with a geojson file I had already created using the code. I just added it. 

AuxiliaryInfo: I made changes to it due to the fact that version_data and signature_data were unused and probably giving me a warning. So I added the underscore to indicate to Rust that the field is intentionally there but unused, basically suppressing the compiler warnings. Wasn't sure if you were ever going to use them, but rather than removing them I just intentionally marked them as unused.

I am no expert with Rust, so many of my changes we the result of error messages when compiling it. My main additions to the code were simply to support the wrapper code as I learned what Rust needed. So, some of the changes are from initial attempts at it, and since it didn't break compilation they remained. 

Greetings from New Zealand.

lvauvillier commented:
Ok, Thanks for your explanation. I agree there are unused variables, and we can address that by simply adding `#[allow(unused)]`.

However, I don't think the PR is mergeable because you aim to export GeoJSON functionalities, whereas I want the bindings to mirror the original library's API—focusing only on parsing, keychain retrieval, and frame/record export.

The CLI crate serves primarily as an example of what can be achieved and how to work with the library.

Nevertheless, thank you for your efforts. This has provided a good starting point for me to understand C bindings, and I definitely plan to support this in the future.

Greetings from Paris.

edancain commented:
Hi Luc,
I agree with you that the PR is not mergeable because of what I want to add as functionality and you keeping it mirroring the original library API. 
I couldn't get their API to build for me, so discovering your comments and your Rust work has been so helpful in my project and I can't thank you enough. 
I'm very glad that what I have shown in the PR will help as a starting point in your understanding of C bindings.
Thank you so much! 
Edan

edancain committed:
Add files via upload

edancain committed:
Updated Files

edancain committed:
Updates to file structure

edancain committed:
Changes to files
